sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RedirectPolicy.java

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.http.policy;
+
+import com.azure.core.http.HttpPipelineCallContext;
+import com.azure.core.http.HttpPipelineNextPolicy;
+import com.azure.core.http.HttpRequest;
+import com.azure.core.http.HttpResponse;
+import reactor.core.publisher.Mono;
+
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * A {@link HttpPipelinePolicy} that redirects a {@link HttpRequest} when an HTTP Redirect is received as
+ * {@link HttpResponse response}.
+ */
+public final class RedirectPolicy implements HttpPipelinePolicy {
+    private final RedirectStrategy redirectStrategy;
+    private final Set<String> attemptedRedirectUrls = new HashSet<>();
+
+    /**
+     * Creates {@link RedirectPolicy} with default {@link DefaultRedirectStrategy} as {@link RedirectStrategy} and
+     * uses the redirect status response code (301, 302, 307, 308) to determine if this request should be redirected.
+     */
+    public RedirectPolicy() {
+        this(new DefaultRedirectStrategy());
+    }
+
+    /**
+     * Creates {@link RedirectPolicy} with the provided {@code redirectStrategy} as {@link RedirectStrategy}
+     * to determine if this request should be redirected.
+     *
+     * @param redirectStrategy The {@link RedirectStrategy} used for redirection.
+     * @throws NullPointerException When {@code redirectStrategy} is {@code null}.
+     */
+    public RedirectPolicy(RedirectStrategy redirectStrategy) {
+        this.redirectStrategy = Objects.requireNonNull(redirectStrategy, "'redirectStrategy' cannot be null.");
+    }
+
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        return attemptRedirect(context, next, context.getHttpRequest(), 1);
+    }
+
+    /**
+     * Function to process through the HTTP Response received in the pipeline
+     * and redirect sending the request with new redirect url.
+     */
+    private Mono<HttpResponse> attemptRedirect(final HttpPipelineCallContext context,
+                                               final HttpPipelineNextPolicy next,
+                                               final HttpRequest originalHttpRequest,
+                                               final int redirectAttempt) {
+        context.setHttpRequest(originalHttpRequest.copy());
+
+        return next.clone().process()
+            .flatMap(httpResponse -> {
+                if (redirectStrategy.shouldAttemptRedirect(context, httpResponse, redirectAttempt,
+                    attemptedRedirectUrls)) {
```

Comment:
I'm a little confused about `attemptedRedirectUrls`. I assumed that this set should be cleared out between redirects, but that doesn't seem to be the case. Clearly I'm misunderstanding something so a comment in the code might be useful.

Reply:
It should be cleared between requests, https://github.com/Azure/azure-sdk-for-java/pull/23617#discussion_r694194464.

Reply:
I don't understand where this is cleared, and I don't think the comment you added helps to clarify things.

Reply:
Updated this to be cleared at per request level https://github.com/Azure/azure-sdk-for-java/pull/23617/files#diff-e8817f59a773c09687266c0d12f83a377a842a5f8a3c9ffcd119ed7857e8ae56R44-R45

Reply:
Thank you - this now works as I expected it to 👍